### ISSUE TITLE
Spark can't read ORC files with numeric column names

### ISSUE THREAD

**Issue by ecurtin** [SparkTC/spark-bench#156]
### Spark-Bench version (version number, tag, or git commit hash)
a1975e52a2167dc3609e615753928ada29a96a8e

### Details of your cluster setup (Spark version, Standalone/Yarn/Local/Etc)
Local[*] 2.2.1 built with Hadoop 2.7.5

### Scala version on your cluster
2.11.8

### Your exact configuration file (with system details anonymized for security)
N/A, custom scratch test script inside repo. See code below.

### Relevant stacktrace
This is from a KMeansDataGen scratch test that I wrote to pinpoint the issue. The code:
```
    val spark = SparkSessionProvider.spark

    val fileNAme = "/tmp/stupid-garbage.orc"

    Utils.deleteRecursively(new File(fileNAme))

    val data = KMeansDataGenerator.generateKMeansRDD(
      spark.sparkContext,
      10,
      2,
      2,
      0.6,
      1
    )

    val schemaString = data.first().indices.map(_.toString).mkString(" ")
    val fields = schemaString.split(" ").map(fieldName => StructField(fieldName, DoubleType, nullable = false))
    val schema = StructType(fields)

    val rowRDD = data.map(arr => Row(arr:_*))
    val df = spark.createDataFrame(rowRDD, schema)

    df.write.orc(fileNAme)

    val readingItIn = spark.read.format("orc").load(fileNAme)

    println(readingItIn.count())

```

And the stacktrace...

```
objc[48187]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/bin/java (0x1061d14c0) and /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/libinstrument.dylib (0x1081f14e0). One of the two will be used. Which one is undefined.
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
18/02/16 14:04:27 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
18/02/16 14:04:31 INFO deprecation: mapred.job.id is deprecated. Instead, use mapreduce.job.id
18/02/16 14:04:31 INFO deprecation: mapred.tip.id is deprecated. Instead, use mapreduce.task.id
18/02/16 14:04:31 INFO deprecation: mapred.task.id is deprecated. Instead, use mapreduce.task.attempt.id
18/02/16 14:04:31 INFO deprecation: mapred.task.is.map is deprecated. Instead, use mapreduce.task.ismap
18/02/16 14:04:31 INFO deprecation: mapred.task.partition is deprecated. Instead, use mapreduce.task.partition
18/02/16 14:04:31 INFO FileOutputCommitter: File Output Committer Algorithm version is 2
18/02/16 14:04:32 INFO FileOutputCommitter: File Output Committer Algorithm version is 2
18/02/16 14:04:32 INFO FileOutputCommitter: Saved output of task 'attempt_20180216140432_0000_m_000000_0' to file:/tmp/stupid-garbage.orc


mismatched input '0' expecting '>'(line 1, pos 7)

== SQL ==
struct<0:double,1:double>
-------^^^

org.apache.spark.sql.catalyst.parser.ParseException: 
mismatched input '0' expecting '>'(line 1, pos 7)

== SQL ==
struct<0:double,1:double>
-------^^^

	at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:197)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:99)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parseDataType(ParseDriver.scala:37)
	at org.apache.spark.sql.hive.orc.OrcFileOperator$$anonfun$readSchema$2.apply(OrcFileOperator.scala:81)
	at org.apache.spark.sql.hive.orc.OrcFileOperator$$anonfun$readSchema$2.apply(OrcFileOperator.scala:77)
	at scala.Option.map(Option.scala:146)
	at org.apache.spark.sql.hive.orc.OrcFileOperator$.readSchema(OrcFileOperator.scala:77)
	at org.apache.spark.sql.hive.orc.OrcFileFormat.inferSchema(OrcFileFormat.scala:58)
	at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$7.apply(DataSource.scala:184)
	at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$7.apply(DataSource.scala:184)
	at scala.Option.orElse(Option.scala:289)
	at org.apache.spark.sql.execution.datasources.DataSource.org$apache$spark$sql$execution$datasources$DataSource$$getOrInferFileFormatSchema(DataSource.scala:183)
	at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:387)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:152)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:135)
	at com.ibm.sparktc.sparkbench.datageneration.KMeansDataGenTest$$anonfun$1.apply$mcV$sp(KMeansDataGenTest.scala:142)
	at com.ibm.sparktc.sparkbench.datageneration.KMeansDataGenTest$$anonfun$1.apply(KMeansDataGenTest.scala:113)
	at com.ibm.sparktc.sparkbench.datageneration.KMeansDataGenTest$$anonfun$1.apply(KMeansDataGenTest.scala:113)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
	at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpec.runTest(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
	at org.scalatest.Suite$class.run(Suite.scala:1147)
	at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
	at com.ibm.sparktc.sparkbench.datageneration.KMeansDataGenTest.org$scalatest$BeforeAndAfterAll$$super$run(KMeansDataGenTest.scala:32)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
	at com.ibm.sparktc.sparkbench.datageneration.KMeansDataGenTest.run(KMeansDataGenTest.scala:32)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1340)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1334)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1334)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1500)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:138)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)


Process finished with exit code 0
```
### Description of your problem and any other relevant info
If instead I use string names, everything runs fine.

```
    val spark = SparkSessionProvider.spark

    val fileNAme = "/tmp/stupid-garbage.orc"

    Utils.deleteRecursively(new File(fileNAme))

    val data = KMeansDataGenerator.generateKMeansRDD(
      spark.sparkContext,
      10,
      2,
      2,
      0.6,
      1
    )

    val schema = StructType( Seq(
      StructField("first", DoubleType, nullable = false),
      StructField("second", DoubleType, nullable = false)
    ))

    val rowRDD = data.map(arr => Row(arr:_*))
    val df = spark.createDataFrame(rowRDD, schema)

    df.write.orc(fileNAme)

    val readingItIn = spark.read.format("orc").load(fileNAme)

    println(readingItIn.count())
```

This is a bug in Spark. Orctools can read this data just fine (note, this is from an earlier run where I used 10 columns instead of 2)

```
$ java -jar orc-tools-1.4.3-uber.jar data /tmp/stupid-garbage.orc/part-00000-b5c7bd91-96da-4ad4-be3e-344f9927f307.snappy.orc
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Processing data file /tmp/stupid-garbage.orc/part-00000-b5c7bd91-96da-4ad4-be3e-344f9927f307.snappy.orc [length: 2130]
{"0":1.8270485047568874,"1":1.4710527183724604,"2":-1.5197866190254334,"3":-1.7711844582389404,"4":0.44956422091644715,"5":1.0953964730122485,"6":-1.307554251837963,"7":-2.2346294442848693,"8":-0.3055112209134002,"9":2.37794142782504}
{"0":0.9347291692457076,"1":0.6301415325622199,"2":0.359320327320848,"3":-0.3041321118604228,"4":-0.5541592545207465,"5":1.7229785119202934,"6":-0.6360722151668361,"7":0.4460981019179567,"8":0.7512720252005798,"9":-0.40247952499145945}
{"0":1.7465164231324293,"1":1.4889575936956514,"2":-0.0503504938463315,"3":-0.303724265255192,"4":0.9384281822779845,"5":1.6625736204947443,"6":0.3893484203521068,"7":-2.2359278852729876,"8":1.8225053639675293,"9":-0.5568787286192368}
{"0":0.8745330692933655,"1":0.5681457474708536,"2":0.025755383462769732,"3":0.6581895996990914,"4":-2.1553239674214884,"5":0.8809879693299965,"6":-0.3837627235107132,"7":0.8362220036118913,"8":-9.817864998528103E-4,"9":0.8810145566600188}
{"0":2.040964654340478,"1":-0.27542822800237676,"2":1.036641159426876,"3":-2.340304518778751,"4":-0.9096197049228927,"5":0.7003437776000738,"6":-0.2051150336764101,"7":-1.4684236963216115,"8":0.6594461906066581,"9":1.7659982932226246}
{"0":2.231509625190634,"1":0.165553950162822,"2":0.4260949945490792,"3":-0.7693854454070543,"4":-0.6010184304571224,"5":-1.5322813419128352,"6":0.41645931177868184,"7":0.943646914808723,"8":1.529473323580286,"9":2.3824894963634837}
{"0":0.7341645256607737,"1":-0.3868184653564506,"2":0.23028151334280977,"3":-1.9924592519362037,"4":-1.133785280324086,"5":1.0236465486339563,"6":-0.45250035600924654,"7":0.8016111499610622,"8":0.27133671356291955,"9":1.7948142028469825}
{"0":0.8479812254247225,"1":-0.6755954007742706,"2":0.4091125897341032,"3":0.5271562961669225,"4":-0.3672388406340187,"5":0.4620552298752151,"6":-0.9780947611105275,"7":-0.35273824155931627,"8":-1.141817762685812,"9":1.1768733350637373}
{"0":2.1483351669714503,"1":1.278789344023917,"2":1.8077303201534083,"3":-0.6180897559364122,"4":0.951903205934471,"5":1.1195476912817746,"6":-0.7280869672421849,"7":-3.0463262655482337,"8":-0.7507643629663004,"9":1.527159737224077}
{"0":1.1873046360362465,"1":0.7992624919449651,"2":-0.02916161495356484,"3":-0.996049204864252,"4":-1.8623112886917839,"5":-0.02624561249499935,"6":-0.6966430197764303,"7":0.1556761234384225,"8":-1.6298216824343283,"9":-0.020073207052344944}
```

And the meta dump:

```
$ java -jar orc-tools-1.4.3-uber.jar meta /tmp/stupid-garbage.orc/part-00000-b5c7bd91-96da-4ad4-be3e-344f9927f307.snappy.orc
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Processing data file /tmp/stupid-garbage.orc/part-00000-b5c7bd91-96da-4ad4-be3e-344f9927f307.snappy.orc [length: 2130]
Structure for /tmp/stupid-garbage.orc/part-00000-b5c7bd91-96da-4ad4-be3e-344f9927f307.snappy.orc
File Version: 0.12 with HIVE_8732
Rows: 10
Compression: SNAPPY
Compression size: 262144
Type: struct<0:double,1:double,2:double,3:double,4:double,5:double,6:double,7:double,8:double,9:double>

Stripe Statistics:
  Stripe 1:
    Column 0: count: 10 hasNull: false
    Column 1: count: 10 hasNull: false min: 0.7341645256607737 max: 2.231509625190634 sum: 14.573087000052697
    Column 2: count: 10 hasNull: false min: -0.6755954007742706 max: 1.4889575936956514 sum: 5.064061284099791
    Column 3: count: 10 hasNull: false min: -1.5197866190254334 max: 1.8077303201534083 sum: 2.6956375601645646
    Column 4: count: 10 hasNull: false min: -2.340304518778751 max: 0.6581895996990914 sum: -7.9099831164112135
    Column 5: count: 10 hasNull: false min: -2.1553239674214884 max: 0.951903205934471 sum: -5.243561157843237
    Column 6: count: 10 hasNull: false min: -1.5322813419128352 max: 1.7229785119202934 sum: 7.109002867740469
    Column 7: count: 10 hasNull: false min: -1.307554251837963 max: 0.41645931177868184 sum: -4.582021596199523
    Column 8: count: 10 hasNull: false min: -3.0463262655482337 max: 0.943646914808723 sum: -6.154791239248962
    Column 9: count: 10 hasNull: false min: -1.6298216824343283 max: 1.8225053639675293 sum: 1.2051368014182793
    Column 10: count: 10 hasNull: false min: -0.5568787286192368 max: 2.3824894963634837 sum: 10.926859588542921

File Statistics:
  Column 0: count: 10 hasNull: false
  Column 1: count: 10 hasNull: false min: 0.7341645256607737 max: 2.231509625190634 sum: 14.573087000052697
  Column 2: count: 10 hasNull: false min: -0.6755954007742706 max: 1.4889575936956514 sum: 5.064061284099791
  Column 3: count: 10 hasNull: false min: -1.5197866190254334 max: 1.8077303201534083 sum: 2.6956375601645646
  Column 4: count: 10 hasNull: false min: -2.340304518778751 max: 0.6581895996990914 sum: -7.9099831164112135
  Column 5: count: 10 hasNull: false min: -2.1553239674214884 max: 0.951903205934471 sum: -5.243561157843237
  Column 6: count: 10 hasNull: false min: -1.5322813419128352 max: 1.7229785119202934 sum: 7.109002867740469
  Column 7: count: 10 hasNull: false min: -1.307554251837963 max: 0.41645931177868184 sum: -4.582021596199523
  Column 8: count: 10 hasNull: false min: -3.0463262655482337 max: 0.943646914808723 sum: -6.154791239248962
  Column 9: count: 10 hasNull: false min: -1.6298216824343283 max: 1.8225053639675293 sum: 1.2051368014182793
  Column 10: count: 10 hasNull: false min: -0.5568787286192368 max: 2.3824894963634837 sum: 10.926859588542921

Stripes:
  Stripe: offset: 3 data: 830 rows: 10 tail: 130 index: 451
    Stream: column 0 section ROW_INDEX start: 3 length 11
    Stream: column 1 section ROW_INDEX start: 14 length 44
    Stream: column 2 section ROW_INDEX start: 58 length 44
    Stream: column 3 section ROW_INDEX start: 102 length 44
    Stream: column 4 section ROW_INDEX start: 146 length 44
    Stream: column 5 section ROW_INDEX start: 190 length 44
    Stream: column 6 section ROW_INDEX start: 234 length 44
    Stream: column 7 section ROW_INDEX start: 278 length 44
    Stream: column 8 section ROW_INDEX start: 322 length 44
    Stream: column 9 section ROW_INDEX start: 366 length 44
    Stream: column 10 section ROW_INDEX start: 410 length 44
    Stream: column 1 section DATA start: 454 length 83
    Stream: column 2 section DATA start: 537 length 83
    Stream: column 3 section DATA start: 620 length 83
    Stream: column 4 section DATA start: 703 length 83
    Stream: column 5 section DATA start: 786 length 83
    Stream: column 6 section DATA start: 869 length 83
    Stream: column 7 section DATA start: 952 length 83
    Stream: column 8 section DATA start: 1035 length 83
    Stream: column 9 section DATA start: 1118 length 83
    Stream: column 10 section DATA start: 1201 length 83
    Encoding column 0: DIRECT
    Encoding column 1: DIRECT
    Encoding column 2: DIRECT
    Encoding column 3: DIRECT
    Encoding column 4: DIRECT
    Encoding column 5: DIRECT
    Encoding column 6: DIRECT
    Encoding column 7: DIRECT
    Encoding column 8: DIRECT
    Encoding column 9: DIRECT
    Encoding column 10: DIRECT

File length: 2130 bytes
Padding length: 0 bytes
Padding ratio: 0%
________________________________________________________________________________________________________________________

```

Because this is a Spark issue, I can't fix the underlying cause. However, this will necessitate a rethink of the strategy for naming columns of RDD's to turn them into Dataframes, like with the KMeans Data Generator which is where I first noticed this issue.

**Comment by ecurtin** [comment]
Edited the report above because I copied the wrong schema creation code that got it working. It's supposed to be:
```
    val schema = StructType( Seq(
      StructField("first", DoubleType, nullable = false),
      StructField("second", DoubleType, nullable = false)
    ))
```

**Comment by ecurtin** [comment]
Closed by #157
